File: aiimpact-demo/cloudflare/src/store.ts
/**
 * D1 access layer.
 *
 * Three operations carry correctness weight and are written as conditional
 * writes rather than read-then-write: claiming a slug, incrementing a
 * participant's generation count, and publishing. All three are contended --
 * 200 people submitting inside the same 30 seconds -- and a check-then-act
 * version of any of them would be wrong under exactly the load this system is
 * built for.
 *
 * D1 is a better fit than the DynamoDB it replaces: the cap becomes an
 * ordinary conditional UPDATE and slug uniqueness an ordinary UNIQUE index,
 * both checked by the engine rather than assembled from condition expressions,
 * with none of the eventual consistency the secondary index had.
 */
import type { Job, JobStatus, Participant } from "./model";

export class NotFound extends Error {}
export class CapReached extends Error {}
export class SlugTaken extends Error {}

function isUniqueViolation(err: unknown): boolean {
  return /UNIQUE constraint failed/i.test(String((err as Error)?.message ?? err));
}

export class Store {
  constructor(private readonly db: D1Database) {}

  // --- participants --------------------------------------------------------

  async participant(code: string): Promise<Participant> {
    const row = await this.db
      .prepare("SELECT * FROM participants WHERE code = ?")
      .bind(code)
      .first<Participant>();
    if (!row) throw new NotFound(`no participant for code`);
    return row;
  }

  /**
   * Marks a code first used and returns the participant.
   *
   * Idempotent by design: a participant who reloads, loses signal mid-request,
   * or reopens the link must not be locked out of their own code with no
   * recovery short of finding a panitia member. Single use is enforced
   * socially -- one code per printed sheet -- not by refusing the second call.
   */
  async redeem(code: string): Promise<Participant> {
    const row = await this.db
      .prepare(
        `UPDATE participants
            SET redeemed_at = COALESCE(redeemed_at, datetime('now'))
          WHERE code = ?
      RETURNING *`,
      )
      .bind(code)
      .first<Participant>();
    if (!row) throw new NotFound(`no participant for code`);
    return row;
  }

  async slugFree(slug: string): Promise<boolean> {
    const row = await this.db
      .prepare("SELECT 1 AS taken FROM participants WHERE slug = ? LIMIT 1")
      .bind(slug)
      .first<{ taken: number }>();
    return row === null;
  }

  /**
   * Binds a slug to a participant, once.
   *
   * `slug IS NULL` makes the binding permanent, which is not cosmetic: a slug
   * is in someone's WhatsApp history the moment it is published, so it must
   * never move to another business. The UNIQUE index catches two participants
   * racing for the same name; the loser retries with the next suffix.
   */
  async claimSlug(code: string, slug: string): Promise<void> {
    let res: D1Result;
    try {
      res = await this.db
        .prepare("UPDATE participants SET slug = ? WHERE code = ? AND slug IS NULL")
        .bind(slug, code)
        .run();
    } catch (err) {
      if (isUniqueViolation(err)) throw new SlugTaken(slug);
      throw err;
    }
    if (res.meta.changes === 0) {
      // Either the code is unknown or a slug is already bound. A participant
      // regenerating already has one, which is the common case and not an error.
      const p = await this.participant(code);
      if (p.slug !== slug) throw new SlugTaken(slug);
    }
  }

  /**
   * Increments the generation count if, and only if, the participant is under
   * the cap, and returns the new count.
   *
   * One statement, not read-check-write. Under the burst this system is
   * designed for, a double tap would otherwise pass a check-then-act test
   * twice -- and this cap is what bounds worst-case spend.
   */
  async countGeneration(code: string, max: number): Promise<number> {
    const row = await this.db
      .prepare(
        `UPDATE participants
            SET generation_count = generation_count + 1
          WHERE code = ? AND generation_count < ?
      RETURNING generation_count`,
      )
      .bind(code, max)
      .first<{ generation_count: number }>();

    if (row) return row.generation_count;

    // No row updated: tell "unknown code" apart from "cap reached" so the
    // participant gets the right message.
    await this.participant(code); // throws NotFound if unknown
    throw new CapReached(code);
  }

  // --- jobs ----------------------------------------------------------------

  async createJob(jobId: string, code: string): Promise<void> {
    await this.db
      .prepare("INSERT INTO jobs (job_id, code, status) VALUES (?, ?, 'queued')")
      .bind(jobId, code)
      .run();
  }

  async job(jobId: string): Promise<Job> {
    const row = await this.db
      .prepare("SELECT * FROM jobs WHERE job_id = ?")
      .bind(jobId)
      .first<Job>();
    if (!row) throw new NotFound(`no job ${jobId}`);
    return row;
  }

  async setJobStatus(
    jobId: string,
    status: JobStatus,
    opts: { url?: string; message?: string } = {},
  ): Promise<void> {
    await this.db
      .prepare(
        `UPDATE jobs
            SET status  = ?,
                url     = COALESCE(?, url),
                message = COALESCE(?, message)
          WHERE job_id = ?`,
      )
      .bind(status, opts.url ?? null, opts.message ?? null, jobId)
      .run();
  }

  // --- pages ---------------------------------------------------------------

  /**
   * Writes the page and marks the job done in a single batch.
   *
   * This atomicity is the reason pages live in D1 rather than object storage:
   * a two-phase write can half-fail, leaving a job marked done with no page
   * behind it -- a participant shown a URL that 404s in front of a customer.
   */
  async publish(args: {
    slug: string;
    html: string;
    jobId: string;
    url: string;
  }): Promise<void> {
    await this.db.batch([
      this.db
        .prepare(
          `INSERT INTO pages (slug, html, updated_at) VALUES (?, ?, datetime('now'))
           ON CONFLICT (slug) DO UPDATE SET html = excluded.html, updated_at = excluded.updated_at`,
        )
        .bind(args.slug, args.html),
      this.db
        .prepare("UPDATE jobs SET status = 'done', url = ?, slug = ? WHERE job_id = ?")
        .bind(args.url, args.slug, args.jobId),
    ]);
  }

  async page(slug: string): Promise<string | null> {
    const row = await this.db
      .prepare("SELECT html FROM pages WHERE slug = ?")
      .bind(slug)
      .first<{ html: string }>();
    return row?.html ?? null;
  }
}

File: aiimpact-demo/terraform/src/internal/store/store.go
// Package store is the DynamoDB access layer.
//
// Two operations here carry correctness weight and are written as conditional
// writes rather than read-then-write: claiming a slug and incrementing a
// participant's generation count. Both are contended -- 200 people submitting
// inside the same 30 seconds -- and a check-then-act version of either would
// be wrong under exactly the load this system is built for.
package store

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	ddbtypes "github.com/aws/aws-sdk-go-v2/service/dynamodb/types"

	"github.com/tampubolon/iaitbjambi.org/aiimpact-demo/internal/model"
)

// Sentinel errors callers branch on. The API turns these into the messages a
// participant actually sees, so they must be distinguishable.
var (
	ErrNotFound   = errors.New("store: not found")
	ErrCapReached = errors.New("store: generation cap reached")
	ErrSlugTaken  = errors.New("store: slug already claimed")
)

// Store holds the table handles. One per process; safe for concurrent use.
type Store struct {
	db           *dynamodb.Client
	jobs         string
	participants string
}

// New builds a Store from the ambient AWS config.
func New(ctx context.Context, jobsTable, participantsTable string) (*Store, error) {
	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		return nil, fmt.Errorf("store: load aws config: %w", err)
	}
	return &Store{
		db:           dynamodb.NewFromConfig(cfg),
		jobs:         jobsTable,
		participants: participantsTable,
	}, nil
}

func key(name, value string) map[string]ddbtypes.AttributeValue {
	return map[string]ddbtypes.AttributeValue{name: &ddbtypes.AttributeValueMemberS{Value: value}}
}

// --- participants ----------------------------------------------------------

// Participant reads one participant by code.
func (s *Store) Participant(ctx context.Context, code string) (*model.Participant, error) {
	out, err := s.db.GetItem(ctx, &dynamodb.GetItemInput{
		TableName: aws.String(s.participants),
		Key:       key("code", code),
	})
	if err != nil {
		return nil, fmt.Errorf("store: get participant: %w", err)
	}
	if out.Item == nil {
		return nil, ErrNotFound
	}

	var p model.Participant
	if err := attributevalue.UnmarshalMap(out.Item, &p); err != nil {
		return nil, fmt.Errorf("store: unmarshal participant: %w", err)
	}
	return &p, nil
}

// Redeem marks a code as first used and returns the participant.
//
// Deliberately idempotent: a participant who reloads the page, loses signal
// mid-request, or reopens the link must not be locked out of their own code
// with no recovery path short of finding a panitia member. Single-use is
// enforced socially -- one code per printed handout -- not by refusing the
// second request.
func (s *Store) Redeem(ctx context.Context, code string) (*model.Participant, error) {
	now, err := attributevalue.Marshal(time.Now().UTC())
	if err != nil {
		return nil, fmt.Errorf("store: marshal time: %w", err)
	}

	out, err := s.db.UpdateItem(ctx, &dynamodb.UpdateItemInput{
		TableName:           aws.String(s.participants),
		Key:                 key("code", code),
		UpdateExpression:    aws.String("SET redeemed_at = if_not_exists(redeemed_at, :now)"),
		ConditionExpression: aws.String("attribute_exists(code)"),
		ExpressionAttributeValues: map[string]ddbtypes.AttributeValue{
			":now": now,
		},
		ReturnValues: ddbtypes.ReturnValueAllNew,
	})
	if err != nil {
		var cond *ddbtypes.ConditionalCheckFailedException
		if errors.As(err, &cond) {
			return nil, ErrNotFound
		}
		return nil, fmt.Errorf("store: redeem: %w", err)
	}

	var p model.Participant
	if err := attributevalue.UnmarshalMap(out.Attributes, &p); err != nil {
		return nil, fmt.Errorf("store: unmarshal participant: %w", err)
	}
	return &p, nil
}

// SlugFree reports whether a label is unclaimed, via the slug GSI.
func (s *Store) SlugFree(ctx context.Context, slug string) (bool, error) {
	out, err := s.db.Query(ctx, &dynamodb.QueryInput{
		TableName:              aws.String(s.participants),
		IndexName:              aws.String("slug-index"),
		KeyConditionExpression: aws.String("slug = :s"),
		ExpressionAttributeValues: map[string]ddbtypes.AttributeValue{
			":s": &ddbtypes.AttributeValueMemberS{Value: slug},
		},
		Limit: aws.Int32(1),
	})
	if err != nil {
		return false, fmt.Errorf("store: query slug: %w", err)
	}
	return len(out.Items) == 0, nil
}

// ClaimSlug binds a slug to a participant, once.
//
// The condition makes this safe against two participants racing for the same
// name: the loser gets ErrSlugTaken and retries with the next suffix. It also
// makes the binding permanent -- a slug is in someone's WhatsApp history the
// moment it is published, so it must never move to a different business.
//
// Note the GSI is eventually consistent, so SlugFree can report a stale free.
// This condition is what actually enforces uniqueness; SlugFree only keeps the
// common case cheap.
func (s *Store) ClaimSlug(ctx context.Context, code, slug string) error {
	_, err := s.db.UpdateItem(ctx, &dynamodb.UpdateItemInput{
		TableName:           aws.String(s.participants),
		Key:                 key("code", code),
		UpdateExpression:    aws.String("SET slug = :slug"),
		ConditionExpression: aws.String("attribute_exists(code) AND attribute_not_exists(slug)"),
		ExpressionAttributeValues: map[string]ddbtypes.AttributeValue{
			":slug": &ddbtypes.AttributeValueMemberS{Value: slug},
		},
	})
	if err != nil {
		var cond *ddbtypes.ConditionalCheckFailedException
		if errors.As(err, &cond) {
			return ErrSlugTaken
		}
		return fmt.Errorf("store: claim slug: %w", err)
	}
	return nil
}

// CountGeneration increments a participant's generation count if, and only if,
// they are under the cap, and returns the new count.
//
// This is a single conditional ADD rather than read-check-write. Under the
// burst this system is designed for, a participant tapping twice would
// otherwise pass the check twice and exceed the cap -- and the cap is what
// bounds worst-case spend.
func (s *Store) CountGeneration(ctx context.Context, code string, max int) (int, error) {
	out, err := s.db.UpdateItem(ctx, &dynamodb.UpdateItemInput{
		TableName:        aws.String(s.participants),
		Key:              key("code", code),
		UpdateExpression: aws.String("SET generation_count = if_not_exists(generation_count, :zero) + :one"),
		ConditionExpression: aws.String(
			"attribute_exists(code) AND (attribute_not_exists(generation_count) OR generation_count < :max)"),
		ExpressionAttributeValues: map[string]ddbtypes.AttributeValue{
			":zero": &ddbtypes.AttributeValueMemberN{Value: "0"},
			":one":  &ddbtypes.AttributeValueMemberN{Value: "1"},
			":max":  &ddbtypes.AttributeValueMemberN{Value: fmt.Sprint(max)},
		},
		ReturnValues: ddbtypes.ReturnValueAllNew,
	})
	if err != nil {
		var cond *ddbtypes.ConditionalCheckFailedException
		if errors.As(err, &cond) {
			// Either the code does not exist or the cap is reached. Tell them
			// apart so the participant gets the right message.
			if _, gErr := s.Participant(ctx, code); errors.Is(gErr, ErrNotFound) {
				return 0, ErrNotFound
			}
			return 0, ErrCapReached
		}
		return 0, fmt.Errorf("store: count generation: %w", err)
	}

	var p model.Participant
	if err := attributevalue.UnmarshalMap(out.Attributes, &p); err != nil {
		return 0, fmt.Errorf("store: unmarshal participant: %w", err)
	}
	return p.GenerationCount, nil
}

// --- jobs ------------------------------------------------------------------

// JobTTL is how long a finished job row survives. Long enough for a
// participant to come back to it during the session, short enough that the
// table empties itself afterwards.
const JobTTL = 7 * 24 * time.Hour

// CreateJob writes a queued job.
func (s *Store) CreateJob(ctx context.Context, j model.Job) error {
	j.CreatedAt = time.Now().UTC()
	j.TTL = j.CreatedAt.Add(JobTTL).Unix()
	j.Status = model.StatusQueued

	item, err := attributevalue.MarshalMap(j)
	if err != nil {
		return fmt.Errorf("store: marshal job: %w", err)
	}
	if _, err := s.db.PutItem(ctx, &dynamodb.PutItemInput{
		TableName: aws.String(s.jobs),
		Item:      item,
	}); err != nil {
		return fmt.Errorf("store: put job: %w", err)
	}
	return nil
}

// Job reads one job.
func (s *Store) Job(ctx context.Context, id string) (*model.Job, error) {
	out, err := s.db.GetItem(ctx, &dynamodb.GetItemInput{
		TableName: aws.String(s.jobs),
		Key:       key("job_id", id),
	})
	if err != nil {
		return nil, fmt.Errorf("store: get job: %w", err)
	}
	if out.Item == nil {
		return nil, ErrNotFound
	}

	var j model.Job
	if err := attributevalue.UnmarshalMap(out.Item, &j); err != nil {
		return nil, fmt.Errorf("store: unmarshal job: %w", err)
	}
	return &j, nil
}

// SetJobStatus advances a job. url and message are optional.
func (s *Store) SetJobStatus(ctx context.Context, id string, st model.JobStatus, url, message string) error {
	values := map[string]ddbtypes.AttributeValue{
		":s": &ddbtypes.AttributeValueMemberS{Value: string(st)},
	}
	expr := "SET #st = :s"
	if url != "" {
		expr += ", #u = :u"
		values[":u"] = &ddbtypes.AttributeValueMemberS{Value: url}
	}
	if message != "" {
		expr += ", #m = :m"
		values[":m"] = &ddbtypes.AttributeValueMemberS{Value: message}
	}

	names := map[string]string{"#st": "status"} // status is a reserved word
	if url != "" {
		names["#u"] = "url"
	}
	if message != "" {
		names["#m"] = "message"
	}

	if _, err := s.db.UpdateItem(ctx, &dynamodb.UpdateItemInput{
		TableName:                 aws.String(s.jobs),
		Key:                       key("job_id", id),
		UpdateExpression:          aws.String(expr),
		ExpressionAttributeNames:  names,
		ExpressionAttributeValues: values,
	}); err != nil {
		return fmt.Errorf("store: set job status: %w", err)
	}
	return nil
}
